### ISSUE TITLE
Bug: React-17.0.0-rc.0 react-reconciler focusedInstanceHandle crash

### ISSUE THREAD

**Issue by drcmda** [facebook/react#19582]
i tried testing the rc with react-three-fiber, it crashed. so tried with a super reduced minimal renderer, here's the link for testing it: https://codesandbox.io/s/reurope-reconciler-forked-wikus?file=/src/App.js:418-492

change the color in line 13 from hotpink to something else and it crashes the reconciler

<img width="735" alt="Screenshot 2020-08-11 at 00 12 46" src="https://user-images.githubusercontent.com/2223602/89836595-69b7f780-db67-11ea-9f20-848b50c5c960.png">

going through the stack trace it passes `focusedInstanceHandle` to `doesFiberContain`, but it is undefined

```jsx
function commitBeforeMutationEffects() {
  ...
        if (doesFiberContain(nextEffect, focusedInstanceHandle)) {
```

this now causes trouble in `doesFiberContain` which checks against null, but not undefined

```jsx
function doesFiberContain(parentFiber, childFiber) {
  var node = childFiber; // undefined
  while (node !== null) {
    ...
    node = node.return;
```

**Comment by gaearon** [comment]
I think if you return `null` from `prepareToCommit` that would fix it.

**Comment by gaearon** [comment]
(Maybe this is a good argument to make the check looser.)

**Comment by drcmda** [comment]
indeed that fixed it! are there any other places it need null?

looks better now in r3f, too. it started to run - for a while - then it wipes out the complete content of a useRef, where i keep my state. will make a new new issue if im certain im not the one causing it.

**Comment by gaearon** [comment]
Not aware of other places. If you narrow it down, file for sure!

**Comment by drcmda** [comment]
@gaearon turns out this was related to https://reactjs.org/blog/2020/08/10/react-v17-rc.html#potential-issues (the new useEffect timing). some even wanted to rely on state, which was disposed earlier.

**Comment by gaearon** [comment]
You can `useLayoutEffect` if being sync is important.

**Comment by HriBB** [comment]
> I think if you return `null` from `prepareToCommit` that would fix it.

That's `prepareForCommit` not `prepareToCommit` :) I just blindly copy-pasted it and was really sad, when it didn't work. Luckily I double-checked the name and saw the problem ... and now it works! Thanks @gaearon.

**Comment by gaearon** [comment]
I added some docs today:

https://github.com/facebook/react/tree/master/packages/react-reconciler#an-incomplete-reference

**Comment by HriBB** [comment]
@gaearon cool thanks! Already bookmarked :)

**Comment by JustinCarmony** [comment]
Yay, this solved a problem for me! Adding some context for future google searchers.

This solved a problem for me while working with building a custom reconciler. I had a stub for `prepareForCommit` that didn't return anything:

```
prepareForCommit: function (rootContainerInstance:Root){}
```

Adding a `return null;` solved it for me:

```
prepareForCommit: function (rootContainerInstance:Root):null { 
        return null;
}
```

```
React will try to recreate this component tree from scratch using the error boundary you provided, ErrorBoundary.
Error caught TypeError: Cannot read property 'return' of undefined
    at doesFiberContain (/Users/jcarmony/.stream-deck/node_modules/react-reconciler/cjs/react-reconciler.development.js:675:17)
    at commitBeforeMutationEffects (/Users/jcarmony/.stream-deck/node_modules/react-reconciler/cjs/react-reconciler.development.js:15864:13)
    at Object.invokeGuardedCallbackProd (/Users/jcarmony/.stream-deck/node_modules/react-reconciler/cjs/react-reconciler.development.js:12101:10)
    at invokeGuardedCallback (/Users/jcarmony/.stream-deck/node_modules/react-reconciler/cjs/react-reconciler.development.js:12292:31)
    at commitRootImpl (/Users/jcarmony/.stream-deck/node_modules/react-reconciler/cjs/react-reconciler.development.js:15652:9)
    at unstable_runWithPriority (/Users/jcarmony/.stream-deck/node_modules/scheduler/cjs/scheduler.development.js:468:12)
    at runWithPriority (/Users/jcarmony/.stream-deck/node_modules/react-reconciler/cjs/react-reconciler.development.js:2495:10)
    at commitRoot (/Users/jcarmony/.stream-deck/node_modules/react-reconciler/cjs/react-reconciler.development.js:15551:3)
    at performSyncWorkOnRoot (/Users/jcarmony/.stream-deck/node_modules/react-reconciler/cjs/react-reconciler.development.js:14857:3)
    at /Users/jcarmony/.stream-deck/node_modules/react-reconciler/cjs/react-reconciler.development.js:2546:26 
    ```

Now, I was using TypeScript, I didn't have time to debug if how it transpiled affected to the issue. But it was a beast to track down and realize. When Step Debugging, it would throw at the `React.useState` function call in a functional component. Trying to think if there are any other details that would be helpful to someone delving into the depths of react reconciler and implementing their own.

**Comment by cliffordfajardo** [comment]
As other have mentioned in the thread, adding null solved my problem for react 17.0.2

Wanted to share an additional resource for folks implementing custom renderers. Below is a link to a 3 part blog series that is pretty in depth & covers the majority of of the hostConfig methods you might encounter
- https://blog.atulr.com/react-custom-renderer-1/
